🎁 📐 Implement all Pythagorean mean metrics, median metrics, and inverse metrics.

cthoyt opened the pull request:
Closes #380. Additionally, implements reporting of the standard deviation, variance, and median absolute deviation of ranks.

cthoyt committed:
Implement inverse geometric mean rank metric

Trigger CI

cthoyt committed:
Update tutorial

Trigger CI

cthoyt committed:
Generalizing rank statistic calculations

cthoyt committed:
Update readme

cthoyt committed:
Update readme

cthoyt committed:
Add remaining implementation

Trigger CI

cthoyt commented on README.md:
```diff
@@ -221,17 +221,26 @@ or the URL for the dataset if neither of the first two are available.
+| Mean Rank (MR)                              | The arithmetic mean over all ranks on, [1, inf). Lower is better.                      |
+| Mean Reciprocal Rank (MRR)                  | The inverse of the harmonic mean over all ranks, on (0, 1]. Higher is better.          |
+| Median Rank                                 | The median over all ranks, on [1, inf). Lower is better.                               |
+| Rank Median Absolute Deviation              | The median absolute deviation over all ranks on, [0, inf). Lower is better.            |
```
thoughts on including MAD, std, and variance in this list?

mberr replied:
I think they are useful, although not something you would e.g. use for early stopping.

cthoyt replied:
I removed these from the home page because I want that to reflect "metrics"

cthoyt commented on src/pykeen/evaluation/rank_based_evaluator.py:
```diff
@@ -27,19 +29,60 @@
+    RANK_PESSIMISTIC: None,  # TODO - research problem
+}
+
+ARITHMETIC_MEAN_RANK = 'arithmetic_mean_rank'  # also known as mean rank (MR)
```
is changing this going to be a problem?

mberr replied:
I guess so, since mean_rank might have been used before. Maybe we can add synonyms for those, similar to what we have e.g. with losses?

cthoyt replied:
I just worked on the automatic canonicalization of the names used, so anyone who uses something like "mean_rank" in their code for HPO or interaction with the results class will be okay.

If they're hard-coding access to the dictionary after export, they'll just have to upgrade. I think it's worth it to make things more descriptive. Counterpoint - this is me being principled over practical.

cthoyt commented on src/pykeen/evaluation/rank_based_evaluator.py:
```diff
@@ -27,19 +29,60 @@
+MEDIAN_RANK = 'median_rank'
+INVERSE_ARITHMETIC_MEAN_RANK = 'inverse_arithmetic_mean_rank'
+INVERSE_GEOMETRIC_MEAN_RANK = 'inverse_geometric_mean_rank'
+INVERSE_HARMONIC_MEAN_RANK = 'inverse_harmonic_mean_rank'  # also known as mean reciprocal rank (MRR)
```
same here

cthoyt committed:
Update default HPO metric

cthoyt committed:
Update rank_based_evaluator.py

cthoyt committed:
Don't include statistics in metrics

This might be a bit misleading

cthoyt committed:
Update CHANGELOG.rst

cthoyt committed:
Re-update nomenclature

Trigger CI

@mberr I added enough backwards compatibility so anyone using the HPO pipeline shouldn't have to modify their code, but the names will now be canonicalized.
